From cf3b6412daad40ef70827776b922e559ab0a5444 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Wed, 11 May 2022 14:35:43 +0200
Subject: [PATCH 198/841] feat(st-bsec): Add support of BSEC_HARDWARE_KEY for
 STM32MP13

Add support of BSEC_HARDWARE_KEY for STM32MP13 when OTPs for "oem_enc_key"
are locked.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git a9a302ce47c95941314cd8f9c2058e3b34631c96 feat(st-bsec): Add support of BSEC_HARDWARE_KEY for STM32MP13]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Ic1428aed98e8f46cb4486dc9279fb335b2766d65
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/251861
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/bsec/bsec2.c         | 27 +++++++++++++++++++++++++++
 plat/st/stm32mp1/stm32mp1_def.h |  2 +-
 2 files changed, 28 insertions(+), 1 deletion(-)

diff --git a/drivers/st/bsec/bsec2.c b/drivers/st/bsec/bsec2.c
index ad70b5ed3..1ec4ea176 100644
--- a/drivers/st/bsec/bsec2.c
+++ b/drivers/st/bsec/bsec2.c
@@ -863,6 +863,9 @@ uint32_t bsec_get_secure_state(void)
 {
 	uint32_t status = bsec_get_status();
 	uint32_t result = BSEC_STATE_INVALID;
+	uint32_t otp_enc_id __unused;
+	uint32_t otp_bit_len __unused;
+	int res __unused;
 
 	if ((status & BSEC_OTP_STATUS_INVALID) != 0U) {
 		result = BSEC_STATE_INVALID;
@@ -879,5 +882,29 @@ uint32_t bsec_get_secure_state(void)
 		}
 	}
 
+#if STM32MP13
+	res = stm32_get_otp_index(ENCKEY_OTP, &otp_enc_id, &otp_bit_len);
+	if ((res == 0) && (otp_bit_len > 0)) {
+		uint32_t start = otp_enc_id / sizeof(uint32_t);
+		size_t otp_nb = round_up(otp_bit_len, sizeof(uint32_t)) /
+				sizeof(uint32_t);
+		size_t otp_locked_nb = 0U;
+		uint32_t idx;
+
+		for (idx = start; idx < start + otp_nb; idx++) {
+			bool locked = false;
+
+			res = bsec_read_sp_lock(idx, &locked);
+			if ((res != 0U) || (locked == 0U)) {
+				break;
+			}
+			otp_locked_nb++;
+		}
+		if (otp_locked_nb == otp_nb) {
+			result |= BSEC_HARDWARE_KEY;
+		}
+	}
+#endif
+
 	return result;
 }
diff --git a/plat/st/stm32mp1/stm32mp1_def.h b/plat/st/stm32mp1/stm32mp1_def.h
index 896826666..b09688a57 100644
--- a/plat/st/stm32mp1/stm32mp1_def.h
+++ b/plat/st/stm32mp1/stm32mp1_def.h
@@ -441,7 +441,7 @@ enum ddr_type {
 #define MONOTONIC_OTP			"monotonic_otp"
 #define UID_OTP				"uid_otp"
 #define PKH_OTP				"pkh_otp"
-#define ENCKEY_OTP			"enckey_otp"
+#define ENCKEY_OTP			"oem_enc_key"
 #define BOARD_ID_OTP			"board_id"
 
 /* OTP mask */
-- 
2.47.2

